lox-gc/src/heap.rs

```diff
@@ -0,0 +1,723 @@
+            bytes_used: Cell::new(0),
+        };
+
+        let ream = heap.space.new_ream(heap.space.available_pages());
+        heap.free_reams.push(ream);
+
+        Some(heap)
+    }
+
+    pub fn bytes_used(&self) -> usize {
+        self.bytes_used.get()
+    }
+
+    pub fn alloc(&self, layout: std::alloc::Layout) -> *mut u8 {
+        let bytes = layout.size();
+        if bytes <= 4096 {
+            let bin_log = bytes.next_power_of_two().trailing_zeros();
+            let class = SizeClass::SMALL[bin_log.saturating_sub(4) as usize]; //TODO constant
+            self.alloc_small(class)
+        } else {
+            self.alloc_ream(bytes_to_pages(bytes) as _)
+        }
+    }
+
+    pub unsafe fn start_gc(&self) {
+        self.space.pds()
+            .for_each(|pd| pd.set_empty());
+    }
+
+    pub unsafe fn mark(&self, ptr: *const u8) {
+        let page = self.space.pd_of(ptr);
+        let offset = ptr.offset_from(page.data(0)) as usize;
+        let index = offset / page.class().block_bytes().unwrap_or(1);
+
+        page.mark(index);
+    }
+
+    pub fn is_marked(&self, ptr: *const u8) -> bool {
+        unsafe {
+            let page = self.space.pd_of(ptr);
+            let offset = ptr.offset_from(page.data(0)) as usize;
+            let index = offset / page.class().block_bytes().unwrap_or(1);
+
+            page.is_marked(index)
+        }
+    }
+
+    pub unsafe fn sweep(&self) {
+        let mut count = 0;
+        for pd in self.space.pds() {
+            count += pd.bytes_used() as usize;
+
+            if pd.class() == SizeClass::Reserved {
+                continue;
+            }
+
+            if pd.is_empty() {
+                pd.unlink();
+                if pd.is_single_page() {
+                    self.free_pages.push(pd);
+                } else {
+                    self.free_reams.push(pd);
+                }
+            } else if pd.is_full() {
+                // Do nothing
+            } else {
+                debug_assert!(pd.class() != SizeClass::Ream);
+                pd.unlink();
+                self.sized_list(pd.class()).push(pd);
+            }
+        }
+        self.bytes_used.set(count);
+    }
+
+    fn sized_list(&self, size_class: SizeClass) -> &PdList {
+        &self.free_sized[size_class as usize - 4] //TODO constant
+    }
+
+    fn take_ream(&self, pages: u32) -> PdRef {
+        for ream in self.free_reams.iter(&self.space) {
+            let (ream, rest) = match ream.split(pages as _) {
+                Some(x) => x,
+                None => continue,
+            };
+
+            match rest {
+                Some(r) if r.is_single_page() => {
+                    self.free_pages.push(r);
+                },
+                Some(r) => {
+                    self.free_reams.push(r);
+                },
+                None => (),
+            };
+
+            ream.unlink();
+            return ream;
+        }
+
+        panic!("No ream of sufficient size available");
```

Comment:
We should use https://doc.rust-lang.org/std/alloc/fn.handle_alloc_error.html instead.